File: core/concurrent/java/fan/concurrent/AtomicIntPeer.java
//
// Copyright (c) 2010, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Feb 10 Brian Frank  Creation
//
package fan.concurrent;

import java.util.concurrent.atomic.AtomicLong;

public final class AtomicIntPeer
{

  public static AtomicIntPeer make(AtomicInt self)
  {
    return new AtomicIntPeer();
  }

  public final long val(AtomicInt self)
  {
    return val.get();
  }

  public final void val(AtomicInt self, long v)
  {
    val.set(v);
  }

  public final long getAndSet(AtomicInt self, long v)
  {
    return val.getAndSet(v);
  }

  public final boolean compareAndSet(AtomicInt self, long expect, long update)
  {
    return val.compareAndSet(expect, update);
  }

  public final long getAndIncrement(AtomicInt self)
  {
    return val.getAndIncrement();
  }

  public final long getAndDecrement(AtomicInt self)
  {
    return val.getAndDecrement();
  }

  public final long getAndAdd(AtomicInt self, long delta)
  {
    return val.getAndAdd(delta);
  }

  public final long incrementAndGet(AtomicInt self)
  {
    return val.incrementAndGet();
  }

  public final long decrementAndGet(AtomicInt self)
  {
    return val.decrementAndGet();
  }

  public final long addAndGet(AtomicInt self, long delta)
  {
    return val.addAndGet(delta);
  }

  public final void increment(AtomicInt self)
  {
    val.incrementAndGet();
  }

  public final void decrement(AtomicInt self)
  {
    val.decrementAndGet();
  }

  public final void add(AtomicInt self, long delta)
  {
    val.addAndGet(delta);
  }

  private final AtomicLong val = new AtomicLong();
}

File: src/concurrent/js/AtomicIntPeer.js
//
// Copyright (c) 2015, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   17 Aug 2015  Matthew Giannini  Creation
//

fan.concurrent.AtomicIntPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.concurrent.AtomicIntPeer.prototype.$ctor = function() {}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.concurrent.AtomicIntPeer.prototype.m_val = 0;
fan.concurrent.AtomicIntPeer.prototype.val = function(self) { return this.m_val; }
fan.concurrent.AtomicIntPeer.prototype.val$ = function(self, val) { this.m_val = val; }

fan.concurrent.AtomicIntPeer.prototype.getAndSet = function(self, val) {
  var old = this.m_val;
  this.m_val = val;
  return old;
}

fan.concurrent.AtomicIntPeer.prototype.compareAndSet = function(self, expect, update) {
  if (this.m_val == expect) {
    this.m_val = update;
    return true;
  }
  return false;
}

fan.concurrent.AtomicIntPeer.prototype.getAndIncrement = function(self) {
  return this.getAndAdd(self, 1);
}

fan.concurrent.AtomicIntPeer.prototype.getAndDecrement = function(self) {
  return this.getAndAdd(self, -1);
}

fan.concurrent.AtomicIntPeer.prototype.getAndAdd = function(self, delta) {
  var old = this.m_val;
  this.m_val = old + delta;
  return old;
}

fan.concurrent.AtomicIntPeer.prototype.incrementAndGet = function(self) {
  return this.addAndGet(self, 1);
}

fan.concurrent.AtomicIntPeer.prototype.decrementAndGet = function(self) {
  return this.addAndGet(self, -1)
}

fan.concurrent.AtomicIntPeer.prototype.addAndGet = function(self, delta) {
  this.m_val = this.m_val + delta;
  return this.m_val;
}

fan.concurrent.AtomicIntPeer.prototype.increment = function(self) {
  this.add(self, 1);
}

fan.concurrent.AtomicIntPeer.prototype.decrement = function(self) {
  this.add(self, -1)
}

fan.concurrent.AtomicIntPeer.prototype.add = function(self, delta) {
  this.m_val = this.m_val + delta;
}